dre-service/src/main/java/io/cdap/re/RulesDB.java

```diff
@@ -147,376 +199,399 @@ public void updateRule(String id, RuleRequest rule) throws RuleNotFoundException
-    Row row = rulebook.get(toKey(rb.getId()));
-    if (!row.isEmpty()) {
+  void createRulebook(RulebookRequest rulebookRequest)
+    throws RulebookAlreadyExistsException, RuleNotFoundException, IOException {
+    Collection<Field<?>> rulebookKeyFields = new ArrayList<>();
+    rulebookKeyFields.add(Fields.stringField(ID_COL, rulebookRequest.getId()));
+
+    Optional<StructuredRow> optionalRulebookStructuredRow = rulebookTable.read(rulebookKeyFields);
+
+    if (optionalRulebookStructuredRow.isPresent()) {
       throw new RulebookAlreadyExistsException(
-        String.format("Rulebook '%s' already exists. Either update it or delete it and create it.", rb.getId())
+        String
+          .format("Rulebook '%s' already exists. Either update it or delete it and create it.", rulebookRequest.getId())
       );
     }
 
-    for (String rule : rb.getRules()) {
-      if(rules.get(toKey(rule)).isEmpty()) {
+    for (String rule : rulebookRequest.getRules()) {
+      Collection<Field<?>> ruleKeyFields = new ArrayList<>();
+      ruleKeyFields.add(Fields.stringField(ID_COL, rule));
+
+      Optional<StructuredRow> optionalRuleStructuredRow = rulesTable.read(ruleKeyFields);
+
+      if (!optionalRuleStructuredRow.isPresent()) {
         throw new RuleNotFoundException(
           String.format("Rulebook '%s' includes a rule '%s' that does not exist. Please add rule first",
-                        rb.getId(), rule)
+                        rulebookRequest.getId(), rule)
         );
       }
     }
 
-    byte[][] columns = new byte[][] {
-      ID, DESCRIPTION, CREATED, UPDATED, USER, SOURCE, VERSION, RULES
-    };
-
-    byte[][] values = new byte[][] {
-      Bytes.toBytes(rb.getId()),
-      Bytes.toBytes(rb.getDescription()),
-      Bytes.toBytes(getCurrentTime()),
-      Bytes.toBytes(getCurrentTime()),
-      Bytes.toBytes(rb.getUser()),
-      Bytes.toBytes(rb.getSource()),
-      Bytes.toBytes(1L),
-      Bytes.toBytes(rb.getRulesString())
-    };
-
-    rulebook.put(toKey(rb.getId()), columns, values);
+    Collection<Field<?>> rulebookFields = new ArrayList<>();
+    rulebookFields.add(Fields.stringField(ID_COL, rulebookRequest.getId()));
+    rulebookFields.add(Fields.stringField(DESCRIPTION_COL, rulebookRequest.getDescription()));
+    rulebookFields.add(Fields.longField(CREATED_COL, getCurrentTime()));
+    rulebookFields.add(Fields.longField(UPDATED_COL, getCurrentTime()));
+    rulebookFields.add(Fields.stringField(USER_COL, rulebookRequest.getUser()));
+    rulebookFields.add(Fields.longField(VERSION_COL, 1L));
+    rulebookFields.add(Fields.stringField(RULES_COL, GSON.toJson(rulebookRequest.getRules())));
+
+    rulebookTable.upsert(rulebookFields);
   }
 
-  public void createRulebook(Rulebook rulebook) throws RulebookAlreadyExistsException {
+  void createRulebook(Rulebook rulebook) throws RulebookAlreadyExistsException, IOException, RuleNotFoundException {
```

Comment:
Can you add some javadocs to this class? I'm not sure why it needs both this method and the method that takes a RulebookRequest.

Reply:
Here is the endpoint where both `createRulebook` methods are called

```
        if (handler.isContentType("application/json")) {
          RulebookRequest rb = gson.fromJson(content, RulebookRequest.class);
          rulesDB.createRulebook(rb);
          id = rb.getId();
        } else if (handler.isContentType("application/rules-engine")) {
          Reader reader = new StringReader(content);
          Compiler compiler = new RulebookCompiler();
          Rulebook rulebook = compiler.compile(reader);
          rulesDB.createRulebook(rulebook);
          id = rulebook.getName();
        } else {
          String header = handler.getHeader(RequestExtractor.CONTENT_TYPE, "");
          ServiceUtils.error(responder, HttpURLConnection.HTTP_BAD_REQUEST, "Unsupported content type " + header + ".");

          return;
        }
```

The first method takes a RulebookRequest, while the second takes a Rulebook. Hence, it is two methods that do the same thing, but with different objects that are passed. I am not sure why you can have two objects to accomplish the same thing. 

What should go in the javadoc besides that naming the objects that are passed to create the rulebook? I don't know what the design decision was to have two separate methods.

Reply:
That's what I'm confused/asking about. Do the methods do anything different? If not they should be consolidated into a single method, with the caller changing a RulebookRequest to a Rulebook or vice versa.

Reply:
I figured out the difference in the logic between the two methods. The `RulebookRequest` method checks if all of the rules in the request exist. Otherwise, it throws an exception. The `Rulebook` method iterates over the rules and attempts to create the rules that have not been created. If there are rules that have been created, the method to create those rules throws an exception, which is caught by this method and ignored (probably not the best approach to implement this functionality. It would be better to check if the rule has not been create it before attempting to create it). I added javadocs describing this difference in functionality.

Reply:
thanks for documenting it. I'm not sure that behavior actually makes any sense from an API perspective, but at least now we know how it behaves.

Reply:
Yes, as I said, there might be a more elegant way of achieving that behavior, or it might not be intended at all. 

Should I document in the code that we might not want to have two separate methods, but rather only one?